### DISCUSSION TITLE
Clipboard popup widget

### DISCUSSION THREAD

**Discussion by salahoued** [LBCrion/sfwbar#190] [Show and tell]
Hi,
After using `rofi dmenu` to get clipboard history for a while, I decided it was time for good `Sfwbar` widget to replace it.
It should be noted that this widget depends on `wl-clipboard` and [`wl-clipboard-history`](https://github.com/janza/wl-clipboard-history) script which in turn depends on `sqlite3`

![20240601_22h22m59s_grim](https://github.com/LBCrion/sfwbar/assets/17504423/93c16432-29c8-4d7b-9a8c-4836e06b0014)

## ~/.config/sfwbar/widgets/clipboard-pop.widget
~~[clipboard-pop.widget.txt](https://github.com/user-attachments/files/15523437/clipboard-pop.widget.txt)~~
[clipboard-pop.widget.txt](https://github.com/user-attachments/files/15535926/clipboard-pop.widget.txt)

## ~/.config/sfwbar/scripts/sfwbar-clipboard.py
~~[sfwbar-clipboard.py.txt](https://github.com/user-attachments/files/15523442/sfwbar-clipboard.py.txt)~~
[sfwbar-clipboard.py.txt](https://github.com/user-attachments/files/15535941/sfwbar-clipboard.py.txt)

UPDATE: when an item in the clipboard is greater than certain length it became impossible to draw the widget, and sometimes causing `sfwbar` to core dump. using `-GtkWidget-ellipsize` to limit the widget width and modifying the python script to not show more than 3 lines in label.

Hope you like it :)

**Comment by 01micko** [comment]
@salahoued nice job but clipboard can be a very complicated thing with all sorts of issues. See https://github.com/labwc/labwc/issues/2042 just to begin going down the rabbit hole!

~I came up with a completely different solution using external apps. I use `wofi` but you could use any other launcher type app or a dedicated clipboard app like `nwg-clipman`.~

EDIT: Sorry didn't read the post properly!

See https://github.com/LBCrion/sfwbar/issues/216#issuecomment-2281601495 for my take on it. Hope it's useful. Cheers!

**Comment by ffs-akl** [comment]
@salahoued  thank you for your widget!   In my case it always shows the following error
```
system("wl-clipboard-history")
    ^^^^^^
NameError: name 'system' is not defined
```
Would you mind helping me see what's wrong? I haven't made any changes.

**Reply by salahoued** [reply]
I'm not sure why.
What distro are you using and which version of python?
can you add at the top of the python script this `from os import system` ?

**Reply by ffs-akl** [reply]
thank you!  With the addition of “form os import system”, the widget can now pop up a window. But it doesn't show what I've copied
![swappy-20241013-203731](https://github.com/user-attachments/assets/a90f9b5d-5714-4d4f-a3dc-719545cba317)

I am using the arch with phython 3.12.7

**Reply by salahoued** [reply]
can you run in the terminal,
`sqlite3 ~/.clipboard.sqlite "select * from c"`
is there any output?

**Reply by ffs-akl** [reply]
no any output message.
![swappy-20241013-204810](https://github.com/user-attachments/assets/5bafed20-ca2e-4763-84c3-ac9fc06c9156)

**Reply by salahoued** [reply]
OK, the database history is empty, there no history in it yet,
You need to add `wl-clipboard-history -t` in you compositor/WM config so it can be started with the compositor start up.
For now you can run it in a terminal or use  your launcher if it support running shell command.

**Reply by salahoued** [reply]
to make sure it is running correctly, do  `pgrep wl-paste` if there is an output, that mean it is watching your wayland clipboard.
then copy something,  then watch if there is any history with `sqlite3 ~/.clipboard.sqlite "select * from c"` or using the `sfwbar` widget

**Reply by ffs-akl** [reply]
I ran `wl-clipboard-history -t ` in terminal,  got the following output message
![swappy-20241013-210854](https://github.com/user-attachments/assets/c22529ef-2e84-4b22-aa84-909916142e29)

edited
I known why.  I needed to let wl-paste know the path to wl-clipboard-list, after adding the path to the function named listen, it worked 
thank you very much!

**Reply by salahoued** [reply]
Is the `wl-clipboard-history` script in your `$PATH`, if not, add it the PATH (copy it to ~/.local/bin) , then try again

That's the output od `wl-paste -w wl-clipboard-history`, because  `wl-clipboard-history` script is not in your path,
copy that script to `~/.local/bin` or `~/bin`

**Comment by salahoued** [comment]
@knm100, It is better to leave `wl-clipboard-history` as is and put/copy it to `~/.local/bin/` directory, that way you can use it as a command shell, and use it whith other tools.

It is also recommended to add it to your compositor config, so every time you log in it will ran automatically
for example:
* for `sway` you can add `exec wl-clipboard-history -t` to your config
* for `hyprland` you can add `exec-once = wl-clipboard-history -t` to you hypr.conf

**Reply by ffs-akl** [reply]
> @knm100, It is better to leave `wl-clipboard-history` as is and put/copy it to `~/.local/bin/` directory, that way you can use it as a command shell, and use it whith other tools.

you are right,  I did so.

Now I can see what I copied in the popup window, but after clicking on the item, I can't paste it with the right mouse button.

**Reply by salahoued** [reply]
Does all the items shown in the widget have the same issue?

**Reply by ffs-akl** [reply]
Sorry for getting back to you late
yes, all of the items have the same issue.

**Reply by salahoued** [reply]
Does `wl-clipboard-history -p $(wl-clipboard-history -l 1 | cut -d',' -f1)` have any output?

**Reply by ffs-akl** [reply]
the output message as the following message
![swappy-20241014-130532](https://github.com/user-attachments/assets/049610b6-aa84-4b39-a2d4-1c17c7ff70e4)

**Reply by salahoued** [reply]
Did you make any changes to `wl-clipboard-history` and the python script?

**Reply by ffs-akl** [reply]
After re-downloading and just copying wl-clipboard-history to $HOME/.local/bin/ and sfwbar-clipboard.py to $HOME/.config/sfwbar/scripts/, without even opening the file, the result is still the same not being able to right-click paste

**Reply by salahoued** [reply]
To be sure, the items are showing correctly on the widget?

Can you try deleting an item?

**Reply by ffs-akl** [reply]
yes,  I sure the items are showing correctly on the widget.  and can delete an item or all item

https://github.com/user-attachments/assets/e43b0a8d-b2fd-4157-bbcd-0f16c9d3213d

**Reply by salahoued** [reply]
In the python script can you try replacing the line 58 in python script
`action[1] = Exec 'bash -c "wl-clipboard-history -p ''' + str(item[0]) + ''' | wl-copy"'`
replace with
```
# action[1] = Exec 'bash -c "wl-clipboard-history -p ''' + str(item[0]) + ''' | wl-copy"'
 action[1] = "sqlite3 ''' + clipboard + ''' 'SELECT contents FROM c WHERE ID = '''+ str(item[0]) + ''' | wl-copy ' "
```

**Reply by ffs-akl** [reply]
it is still the same not being able to right-click paste

edited 
Thanks for your help, please see the following video. maybe it is helpful.

https://github.com/user-attachments/assets/b95dcecc-7645-4e18-beae-47bc1eb5c757

**Reply by salahoued** [reply]
Can you run the python script `sfwbar-clipboard.py` in the terminal and paste the output here?

**Reply by ffs-akl** [reply]
Please see the following output message

[knm@archlinux ~]$ $HOME/.config/sfwbar/scripts/sfwbar-clipboard.py
```
PopUp "ClipboardPopup" {
  AutoClose = true
  style = "clip_pop_style"

  grid "ClipboardPopupGrid" {
   style = "clip_pop_grid_style"
   #css = "* { -GtkWidget-direction: bottom; -GtkWidget-max-width: 400px; - }"

   grid "ClipboardHeder" {
     style = "clip_header_style"

     label "count_label" {
       value = "1 Items present"
       css = "* { -GtkWidget-align: 0;  -GtkWidget-hexpand: true; }"
     }
     button "clipitem_clearall" {
       style = "clip_item_button_style"
       value = "edit-delete-shred"
       action[1] = 'sqlite3 /home/knm/.clipboard.sqlite "DELETE FROM c"'
       tooltip = "Clear All items in clipboard history"
     }
    }

    grid {
        style = "clip_item_style"
        label "clip_item_label" {
            style = "clip_item_label_style"
            value = "sfwbar-clipboard.py"
            #action[1] = Exec 'bash -c "wl-clipboard-history -p 118 | wl-copy"'
            action[1] = "sqlite3 /home/knm/.clipboard.sqlite 'SELECT contents FROM c WHERE ID = 118 | wl-copy ' "
            tooltip = "Click on item to copy it to clipboard."
        }
        button "clip_item_button" {
            style = "clip_item_button_style"
            value = "edit-delete"
            action[1] = "sqlite3 /home/knm/.clipboard.sqlite 'DELETE FROM c WHERE ID = 118'"
            tooltip = "Click to remove the item from clipboard history."
        }
        }
    
}
 }

```

**Reply by salahoued** [reply]
Sorry, I gave you the wrong command earlier, in `sfwbar-clipboard.py` make it like this:
`action[1] = Exec "sqlite3 ''' + clipboard + ''' 'SELECT contents FROM c WHERE ID = ''' + str(item[0]) + ''' ' | wl-copy"`

**Reply by ffs-akl** [reply]
it is still the same not being able to right-click paste
Please see the following output message (run the python script sfwbar-clipboard.py in the terminal)


```
PopUp "ClipboardPopup" {
  AutoClose = true
  style = "clip_pop_style"

  grid "ClipboardPopupGrid" {
   style = "clip_pop_grid_style"
   #css = "* { -GtkWidget-direction: bottom; -GtkWidget-max-width: 400px; - }"

   grid "ClipboardHeder" {
     style = "clip_header_style"

     label "count_label" {
       value = "3 Items present"
       css = "* { -GtkWidget-align: 0;  -GtkWidget-hexpand: true; }"
     }
     button "clipitem_clearall" {
       style = "clip_item_button_style"
       value = "edit-delete-shred"
       action[1] = 'sqlite3 /home/knm/.clipboard.sqlite "DELETE FROM c"'
       tooltip = "Clear All items in clipboard history"
     }
    }

    grid {
        style = "clip_item_style"
        label "clip_item_label" {
            style = "clip_item_label_style"
            value = "it is still the same not being able to right-click paste"
            action[1] = "sqlite3 /home/knm/.clipboard.sqlite 'SELECT contents FROM c WHERE ID = 123 ' | wl-copy"
            tooltip = "Click on item to copy it to clipboard."
        }
        button "clip_item_button" {
            style = "clip_item_button_style"
            value = "edit-delete"
            action[1] = "sqlite3 /home/knm/.clipboard.sqlite 'DELETE FROM c WHERE ID = 123'"
            tooltip = "Click to remove the item from clipboard history."
        }
        }
    
    grid {
        style = "clip_item_style"
        label "clip_item_label" {
            style = "clip_item_label_style"
            value = "action[1] = \"sqlite3 \'\'\' + clipboard + \'\'\' \'SELECT contents FROM c WHERE ID = \'\'\' + str(item[0]) + \'\'\' \' | wl-copy\""
            action[1] = "sqlite3 /home/knm/.clipboard.sqlite 'SELECT contents FROM c WHERE ID = 122 ' | wl-copy"
            tooltip = "Click on item to copy it to clipboard."
        }
        button "clip_item_button" {
            style = "clip_item_button_style"
            value = "edit-delete"
            action[1] = "sqlite3 /home/knm/.clipboard.sqlite 'DELETE FROM c WHERE ID = 122'"
            tooltip = "Click to remove the item from clipboard history."
        }
        }
    
    grid {
        style = "clip_item_style"
        label "clip_item_label" {
            style = "clip_item_label_style"
            value = "knm100/lxqt-policykit"
            action[1] = "sqlite3 /home/knm/.clipboard.sqlite 'SELECT contents FROM c WHERE ID = 121 ' | wl-copy"
            tooltip = "Click on item to copy it to clipboard."
        }
        button "clip_item_button" {
            style = "clip_item_button_style"
            value = "edit-delete"
            action[1] = "sqlite3 /home/knm/.clipboard.sqlite 'DELETE FROM c WHERE ID = 121'"
            tooltip = "Click to remove the item from clipboard history."
        }
        }
    
}
 }

```

**Reply by salahoued** [reply]
These are the commands that I tried and are "working":

`action[1] = Exec 'bash -c "wl-clipboard-history -p ''' + str(item[0]) + ''' | wl-copy"'`
OR

`action[1] = Exec 'bash -c "wl-copy $(sqlite3 ''' + clipboard + ''' \\"SELECT contents FROM c WHERE ID = ''' + str(item[0]) + ''' \\")" '`

OR

`action[1] = Exec 'bash -c "wl-copy $(wl-clipboard-history -p ''' + str(item[0]) + ''')"'`

I still don't know the reason, why its not working for you (Even the first isn't).

**Reply by salahoued** [reply]
@knm100, if nothing is working for you can try the wrappers under the `contrib` directory in [wl-clipboard-history](https://github.com/janza/wl-clipboard-history) repository,
the `dmenu-wrapper` can be modified to be used with any other tools e.g. `rofi` or `wofi`
instead of
`wl-clipboard-history -p "$(wl-clipboard-history -l 20 | dmenu | cut -d ',' -f1)" | wl-copy`
 
you can put for `rofi`
`wl-clipboard-history -p "$(wl-clipboard-history -l 20 | rofi -dmenu | cut -d ',' -f1)" | wl-copy`
or for `wofi`
`wl-clipboard-history -p "$(wl-clipboard-history -l 20 | wofi --dmenu | cut -d ',' -f1)" | wl-copy`

**Reply by ffs-akl** [reply]
Thank you for help.   I will try it.
